diff -ruN a/kernel/supercalls.c b/kernel/supercalls.c
--- a/kernel/supercalls.c
+++ b/kernel/supercalls.c
@@ -1051,6 +1051,12 @@
 			return 0;
 		}
 #endif // #ifdef CONFIG_KSU_SUSFS_SUS_KSTAT
+#ifdef CONFIG_KSU_SUSFS_SUS_KSTAT_REDIRECT
+		if (cmd == CMD_SUSFS_ADD_SUS_KSTAT_REDIRECT) {
+			susfs_add_sus_kstat_redirect(arg);
+			return 0;
+		}
+#endif
 #ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
 		if (cmd == CMD_SUSFS_SET_UNAME) {
 			susfs_set_uname(arg);
